File: backend/routers/lookup.py
"""Lookup endpoints — dropdown/auto-complete sources for the admin UI.

These return small reference lists (corporations, regions, facility types,
clients, partners). All public reads, no auth.
"""
from fastapi import APIRouter

from supabase_client import supabase
from constants import ORDER_TYPES

router = APIRouter()


@router.get("/api/lookup/corporations")
def lookup_corporations():
    r = supabase.schema("pmis").from_("corporation").select("id,name,code").order("id").execute()
    return r.data or []


@router.get("/api/lookup/regions")
def lookup_regions():
    r = supabase.schema("pmis").from_("region_code").select("code,name,region_group").order("code").execute()
    return r.data or []


@router.get("/api/lookup/facility-types")
def lookup_facility_types():
    r = supabase.schema("pmis").from_("facility_type").select("code,name,division").order("name").execute()
    rows = r.data or []
    # facility_type 테이블에 발주유형(BTL/CMR/민간 등)이 섞여 들어와 있어 제거
    return [row for row in rows if (row.get("name") or "") not in ORDER_TYPES]


@router.get("/api/lookup/order-types")
def lookup_order_types():
    """발주유형은 lookup 테이블이 없고 ORDER_TYPES 상수로 관리되므로 그대로 노출."""
    return sorted(ORDER_TYPES)


@router.get("/api/lookup/clients")
def lookup_clients():
    r = supabase.schema("pmis").from_("client_org").select("id,name,org_type").order("name").execute()
    return r.data or []


@router.get("/api/lookup/partners")
def lookup_partners():
    """JV 파트너(하부업체) 목록 — 자동완성용."""
    r = supabase.schema("pmis").from_("partner_company").select("id,name,is_group_member").order("name").execute()
    return r.data or []

File: web/src/lib/api/lookup.ts
import { createClient } from "@/lib/supabase/client";

/* ── Lookup response types — shared with site-form-dialog ── */

export interface Corporation { id: number; name: string; code: string }
export interface Region { code: string; name: string; region_group: string | null }
export interface FacilityType { code: string; name: string; division: string | null }
export interface ClientOrg { id: number; name: string; org_type: string | null }
export interface PartnerCompany { id: number; name: string; is_group_member: boolean }

/** Reference-table reads. These tables are protected by RLS — only an
 *  approved authenticated user can SELECT (see migration 012). The
 *  browser Supabase client auto-attaches the user's JWT, so no manual
 *  auth header juggling here.
 *
 *  We deliberately bypass the Python backend for these reads: the
 *  payload is small reference data and skipping the extra Next.js → Python
 *  hop saves 10–30 ms per call. Mutations to these tables (and any
 *  read that needs server-side caching or business logic) still go
 *  through the backend. */

const ORDER_TYPES = ["BTL", "CMR", "공공", "민간", "외주"];

export async function fetchCorporations(): Promise<Corporation[]> {
  const supabase = createClient();
  const { data, error } = await supabase
    .schema("pmis")
    .from("corporation")
    .select("id,name,code")
    .order("id");
  if (error) throw error;
  return (data ?? []) as Corporation[];
}

export async function fetchRegions(): Promise<Region[]> {
  const supabase = createClient();
  const { data, error } = await supabase
    .schema("pmis")
    .from("region_code")
    .select("code,name,region_group")
    .order("code");
  if (error) throw error;
  return (data ?? []) as Region[];
}

export async function fetchFacilityTypes(): Promise<FacilityType[]> {
  const supabase = createClient();
  const { data, error } = await supabase
    .schema("pmis")
    .from("facility_type")
    .select("code,name,division")
    .order("name");
  if (error) throw error;
  // facility_type 테이블에 발주유형(BTL/CMR/민간 등)이 섞여 들어와 있어 제거
  // (백엔드 라우터가 하던 필터를 클라이언트로 옮김 — 양쪽이 일치해야 한다)
  return ((data ?? []) as FacilityType[]).filter((row) => !ORDER_TYPES.includes(row.name ?? ""));
}

export async function fetchClients(): Promise<ClientOrg[]> {
  const supabase = createClient();
  const { data, error } = await supabase
    .schema("pmis")
    .from("client_org")
    .select("id,name,org_type")
    .order("name");
  if (error) throw error;
  return (data ?? []) as ClientOrg[];
}

export async function fetchOrderTypes(): Promise<string[]> {
  // 발주유형은 lookup 테이블이 없고 ORDER_TYPES 상수로 관리되므로 그대로 노출.
  return Promise.resolve([...ORDER_TYPES].sort());
}

export async function fetchPartners(): Promise<PartnerCompany[]> {
  const supabase = createClient();
  const { data, error } = await supabase
    .schema("pmis")
    .from("partner_company")
    .select("id,name,is_group_member")
    .order("name");
  if (error) throw error;
  return (data ?? []) as PartnerCompany[];
}

/** Convenience: fetch every lookup in parallel for forms that need them all
 *  (used by site-form-dialog). Returns a tuple in a stable order. */
export async function fetchAllLookups(): Promise<
  [Corporation[], Region[], FacilityType[], ClientOrg[], string[], PartnerCompany[]]
> {
  return Promise.all([
    fetchCorporations(),
    fetchRegions(),
    fetchFacilityTypes(),
    fetchClients(),
    fetchOrderTypes(),
    fetchPartners(),
  ]);
}
